x-pack/apm-server/sampling/groups.go

```diff
@@ -16,41 +16,99 @@ import (
+)
 
 // traceGroups maintains a collection of trace groups.
 type traceGroups struct {
-	// defaultSamplingFraction holds the default sampling fraction to
-	// set for new trace groups. See traceGroup.samplingFraction.
-	defaultSamplingFraction float64
-
 	// ingestRateDecayFactor is λ, the decay factor used for calculating the
 	// exponentially weighted moving average ingest rate for each trace group.
 	ingestRateDecayFactor float64
 
-	// maxGroups holds the maximum number of trace groups to maintain.
-	// Once this is reached, no new trace group events will be sampled.
-	maxGroups int
+	// maxDynamicServices holds the maximum number of dynamic service groups
+	// to maintain. Once this is reached, no new dynamic service groups will
+	// be created, and events may be dropped.
+	maxDynamicServices int
+
+	// catchallServicePolicies, if non-nil, holds policies that apply to
+	// services which are not explicitly specified.
+	catchallServicePolicies []Policy
+
+	mu            sync.RWMutex
+	staticGroups  map[string]serviceGroups
+	dynamicGroups map[string]serviceGroups
+}
+
+type serviceGroupKey struct {
+	serviceEnvironment string
+	traceOutcome       string
+	traceName          string
+}
+
+func (k *serviceGroupKey) match(tx *model.Transaction) bool {
+	if k.serviceEnvironment != "" && k.serviceEnvironment != tx.Metadata.Service.Environment {
+		return false
+	}
+	if k.traceOutcome != "" && k.traceOutcome != tx.Outcome {
+		return false
+	}
+	if k.traceName != "" && k.traceName != tx.Name {
+		return false
+	}
+	return true
+}
+
+type serviceGroups []serviceGroup
 
-	mu sync.RWMutex
+type serviceGroup struct {
+	key serviceGroupKey
+	g   *traceGroup
+}
 
-	// TODO(axw) enable the user to define specific trace groups, along with
-	// a specific sampling rate. The "groups" field would then track all other
-	// trace groups, up to a configured maximum, using the default sampling rate.
-	groups map[traceGroupKey]*traceGroup
+// get returns the traceGroup to which tx should be added based on the
+// defined sampling policies, matching policies in the order given.
+func (sgs serviceGroups) get(tx *model.Transaction) (*traceGroup, bool) {
+	for _, sg := range sgs {
+		if sg.key.match(tx) {
+			return sg.g, true
+		}
+	}
+	return nil, false
 }
 
 func newTraceGroups(
-	maxGroups int,
-	defaultSamplingFraction float64,
+	policies []Policy,
+	maxDynamicServices int,
 	ingestRateDecayFactor float64,
 ) *traceGroups {
-	return &traceGroups{
-		defaultSamplingFraction: defaultSamplingFraction,
-		ingestRateDecayFactor:   ingestRateDecayFactor,
-		maxGroups:               maxGroups,
-		groups:                  make(map[traceGroupKey]*traceGroup),
+	groups := &traceGroups{
+		ingestRateDecayFactor: ingestRateDecayFactor,
+		maxDynamicServices:    maxDynamicServices,
+		staticGroups:          make(map[string]serviceGroups),
+		dynamicGroups:         make(map[string]serviceGroups),
+	}
+	for _, policy := range policies {
+		if policy.ServiceName == "" {
```

Comment:
why empty service name is a special case, can't this logic be added to `match`?

Reply:
That would be technically possible, but this approach:

enables the invariant that a serviceGroup (serviceGroupKey) is always associated with a specific service: either one defined statically (in the policy criteria) or dynamically (no service defined in criteria). This invariant is useful to ensure we can't group multiple services for the purposes of sampling (particularly the weighted random sampling bit), which would rarely if ever make much sense.
simplifies keeping track of how many "dynamic service groups" have been created, i.e. unique groups matching a policy without an explicit service name specified. this is used to enforce the max groups config.

Reply:
Ok, maybe put that on a comment :)